netcanv-relay/src/main.rs

```diff
@@ -338,24 +333,48 @@ async fn handle_packet(
 }
 
 async fn read_packets(
-   mut read: OwnedReadHalf,
-   write: Arc<Mutex<OwnedWriteHalf>>,
+   mut read: Stream,
+   write: Arc<Mutex<Sink>>,
    address: SocketAddr,
    state: &Mutex<State>,
 ) -> anyhow::Result<()> {
-   loop {
-      // This is a bit of a workaround because bincode can't read from async streams.
-      let packet: Packet = {
-         let packet_size = read.read_u32().await?;
-         if packet_size > relay::MAX_PACKET_SIZE {
-            anyhow::bail!("packet is too big");
+   while let Some(message) = read.next().await {
+      match message {
+         Ok(Message::Binary(buffer)) => {
+            if buffer.len() > relay::MAX_PACKET_SIZE as usize {
+               anyhow::bail!("packet is too big");
+            }
+
+            let packet = bincode::deserialize(&buffer)?;
+            handle_packet(&write, address, &state, packet).await?;
+         }
+         Ok(Message::Close(frame)) => {
+            if let Some(frame) = frame {
+               log::info!("client disconnected, reason: {}", frame.reason);
+               return Ok(());
+            }
          }
-         let mut buffer = vec![0; packet_size as usize];
-         read.read_exact(&mut buffer).await?;
-         bincode::deserialize(&buffer)?
-      };
-      handle_packet(&write, address, &state, packet).await?;
+         Ok(_) => log::info!("got ignored message"),
+         Err(e) => {
+            use tungstenite::Error::*;
+            match e {
+               ConnectionClosed => {
+                  break;
+               }
+               AlreadyClosed => {
+                  // According to the documentation this error is the fault of the programmer.
+                  // However, this error would crash the entire matchmaker and *all* rooms,
+                  // so it's better to treat it as a simple error and end the connection.
+                  log::error!("the connection has been closed, but the relay is trying to work with already closed connection.");
```

Comment:
I don't understand what this error message means, it could possibly be reworded to be more clear.

Reply:
I don't know how it should be reworded either. This error is ok for me, but that's probably because I understand where the error is coming from.